### PR TITLE
#813 adding the Distribution class and a corresponding catalog

### PR DESCRIPTION
**Pull request by nimshi89**
this branch is part of https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gh/gss-cogs/csvcubed/813

### CODE REVIEW

File: src/csvcubedmodels/rdf/dcat.py [hunk 1]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
+        str,
+        Triple(
+            DCAT.temporalResolution,
+            PropertyStatus.optional,
+            lambda l: Literal(l, XSD.duration),
+        ),
+    ]
+
+    conforms_to: Ann[str, Triple(DCTERMS.conformsTo, PropertyStatus.recommended, URIRef)]
+    description: Ann[str, Triple(DCTERMS.description, PropertyStatus.recommended, Literal)]
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: Shouldn't this use `map_str_to_en_literal` again?

File: src/csvcubedmodels/rdf/dcat.py [hunk 2]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
         self.rdf_types.add(DCAT.Dataset)
 
+#The new class would be created here which also inherits the Resource class
+class Distribution(Resource):
+
+    isDistributionOf: Ann[
+        str, Triple(DCAT.isDistributionOf, PropertyStatus.recommended, URIRef)
+    ]
+    #these two will need changing probably
+    access_service: Ann[str, Triple(DCAT.accessService, PropertyStatus.mandatory, URIRef)]
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: I don't think any of the `mandatory` triples here should actually be mandatory. I think `title` should be mandatory but that's about it.

File: src/csvcubedmodels/rdf/dcat.py [hunk 3]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
         Resource.__init__(self, uri)
         self.rdf_types.add(DCAT.Dataset)
 
+#The new class would be created here which also inherits the Resource class
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: You probably want to remove this.

File: src/csvcubedmodels/rdf/dcat.py [hunk 4]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
+class Distribution(Resource):
+
+    isDistributionOf: Ann[
+        str, Triple(DCAT.isDistributionOf, PropertyStatus.recommended, URIRef)
+    ]
+    #these two will need changing probably
+    access_service: Ann[str, Triple(DCAT.accessService, PropertyStatus.mandatory, URIRef)]
+    access_url: Ann[str, Triple(DCAT.accessURL, PropertyStatus.mandatory, URIRef)]
+
+    byte_size: Ann[str, Triple(DCAT.byteSize, PropertyStatus.mandatory, Literal)]
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: `byte_size` doesn't have the correct type of literal. Take another look at the DCATv2 standard.

File: src/csvcubedmodels/rdf/dcat.py [hunk 5]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
+    compress_format: Ann[str, Triple(DCAT.compressFormat, PropertyStatus.recommended, URIRef)]
+    download_url: Ann[str, Triple(DCAT.downloadURL, PropertyStatus.recommended, URIRef)]
+    media_type: Ann[str, Triple(DCAT.mediaType, PropertyStatus.recommended, URIRef)]
+
+    #not sure about this
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: Well, it's definitely a URIRef, but I'd leave it as optional.

File: src/csvcubedmodels/rdf/dcat.py [hunk 6]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
+    temporal_resolution: Ann[
+        str,
+        Triple(
+            DCAT.temporalResolution,
+            PropertyStatus.optional,
+            lambda l: Literal(l, XSD.duration),
+        ),
+    ]
+
+    conforms_to: Ann[str, Triple(DCTERMS.conformsTo, PropertyStatus.recommended, URIRef)]
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: I'd leave this as optional, not recommended.

File: src/csvcubedmodels/rdf/dcat.py [hunk 7]
```diff
@@ -119,6 +119,60 @@ def __init__(self, uri: str):
...
+    description: Ann[str, Triple(DCTERMS.description, PropertyStatus.recommended, Literal)]
+
+    #the formt keyword might be an issue
+    format: Ann[str, Triple(DCTERMS.format, PropertyStatus.recommended, URIRef)]
+    issued: Ann[str, Triple(DCTERMS.issued, PropertyStatus.recommended, URIRef)]
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: `issued` and `modified` should be date times shouldn't they?

File: src/csvcubedmodels/rdf/dcat.py [hunk 8]
```diff
@@ -165,3 +219,24 @@ def __init__(self, uri: str):
         self.rdf_types.add(DCAT.Catalog)
 
         self.records = set()
+
+class Catalog(Distribution):
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: Why have you chosen to include `Catalog` here? I don't believe we need it, do we?

It certainly shouldn't inherit from `Distribution`.

File: src/csvcubedmodels/rdf/dcat.py (hunk 1 above)
Comment (nimshi89) [reply src/csvcubedmodels/rdf/dcat.py]: I thought it is using a literal (as I took it form the documentation)

Comment (nimshi89) [reply src/csvcubedmodels/rdf/dcat.py]: but yeah I see why it's not correct this way

File: src/csvcubedmodels/rdf/dcat.py (hunk 2 above)
Comment (nimshi89) [reply src/csvcubedmodels/rdf/dcat.py]: okay I kept the title mandatory and the rest is recommended

File: src/csvcubedmodels/rdf/dcat.py (hunk 3 above)
Comment (nimshi89) [reply src/csvcubedmodels/rdf/dcat.py]: done

File: src/csvcubedmodels/rdf/dcat.py (hunk 4 above)
Comment (nimshi89) [reply src/csvcubedmodels/rdf/dcat.py]: sorted

File: src/csvcubedmodels/rdf/dcat.py [hunk 9]
```diff
@@ -119,6 +119,54 @@ def __init__(self, uri: str):
         Resource.__init__(self, uri)
         self.rdf_types.add(DCAT.Dataset)
 
+class Distribution(Resource):
+
```
Comment (roblinksdata) [inline src/csvcubedmodels/rdf/dcat.py]: Should `Distribution` really inherit from `Resource`? Does the diagram show this?


The diagram

Comment (roblinksdata) [reply src/csvcubedmodels/rdf/dcat.py]: It shouldn't inherit from `dcat:Dataset` either. That arrow is showing a `hasA` relationship, not an inheritance relationship

File: sonar-project.properties [hunk 10]
```diff
@@ -3,6 +3,8 @@ sonar.organization=gss-cogs
 
 sonar.python.version=3.9, 3.10
 
+sonar.exclusions=src/csvcubedmodels/rdf/dcat.py
```
Comment (roblinksdata) [inline sonar-project.properties]: I was expecting this to be in the online config temporarily whilst we do this PR. Can you remove it from here please?

Comment (roblinksdata) [reply sonar-project.properties]: image

**Review by roblinksdata — APPROVED** [review]
image

**Pull request merged**